### PR TITLE
JavaScript refactor of OutgoingConnectionFactory, LocatorInfo, and RouterInfo

### PR DESCRIPTION
**Pull request by pepone**
This PR simplifies the implementation of OutgoingConnectionFactory, LocatorInfo, and RouterInfo in JavaScript.

### CODE REVIEW

File: js/src/Ice/ConnectionI.js [hunk 1]
```diff
@@ -218,15 +218,6 @@ export class ConnectionI {
         return this._state > StateNotValidated && this._state < StateClosing;
     }
 
-    isFinished() {
```
Comment (pepone) [inline js/src/Ice/ConnectionI.js:221]: Unused code

File: js/src/Ice/LocatorInfo.js [hunk 2]
```diff
@@ -50,24 +49,6 @@ export class LocatorInfo {
         return this._locator;
     }
 
-    getLocatorRegistry() {
```
Comment (pepone) [inline js/src/Ice/LocatorInfo.js:53]: Unused code

File: js/src/Ice/LocatorInfo.js [hunk 3]
```diff
@@ -349,63 +330,47 @@ class RequestCallback {
         this._promise = promise;
     }
 
-    response(locatorInfo, proxy) {
+    async response(locatorInfo, proxy) {
```
Comment (pepone) [inline js/src/Ice/LocatorInfo.js:333]: Updated to use async/await

File: js/src/Ice/ObjectAdapter.js [hunk 4]
```diff
@@ -8,7 +8,6 @@ const { Identity } = Ice_Identity;
 import { generateUUID } from "./UUID.js";
 import {
     AlreadyRegisteredException,
-    FeatureNotSupportedException,
```
Comment (pepone) [inline js/src/Ice/ObjectAdapter.js:11]: Unused import

File: js/src/Ice/OutgoingConnectionFactory.js [hunk 5]
```diff
@@ -200,45 +200,18 @@ export class OutgoingConnectionFactory {
             throw new CommunicatorDestroyedException();
         }
 
-        //
-        // Try to get the connection.
-        //
-        while (true) {
```
Comment (pepone) [inline js/src/Ice/OutgoingConnectionFactory.js:206]: We don't need a loop here, we are only doing 1 interation.

File: js/src/Ice/RouterInfo.js [hunk 6]
```diff
@@ -47,41 +46,35 @@ export class RouterInfo {
...
-                .then(result => this.setClientEndpoints(result[0], result[1] !== undefined ? result[1] : true, promise))
-                .catch(promise.reject);
+    async getClientEndpoints() {
+        if (this._clientEndpoints === null) {
+            const [clientProxy, hasRoutingTable] = await this._router.getClientProxy();
+            this._hasRoutingTable = hasRoutingTable ?? true;
+            this._clientEndpoints =
+                clientProxy === null
+                    ? this._router._getReference().getEndpoints()
+                    : clientProxy._getReference().getEndpoints();
```
Comment (pepone) [inline js/src/Ice/RouterInfo.js:56]: Move the implementation of `setClientEndpoints` here, as it is only used in this place.

File: js/src/Ice/OutgoingConnectionFactory.js (hunk 5 above)
Comment (externl) [reply js/src/Ice/OutgoingConnectionFactory.js:206]: We should fix in Java as well

**Review by externl — APPROVED** [review]

File: js/src/Ice/RouterInfo.js [hunk 7]
```diff
@@ -47,41 +46,35 @@ export class RouterInfo {
...
-            }
-            serverProxy = serverProxy.ice_router(null); // The server proxy cannot be routed.
-            return serverProxy._getReference().getEndpoints();
-        });
+    async getServerEndpoints() {
+        let serverProxy = await this._router.getServerProxy();
+        if (serverProxy === null) {
+            throw new NoEndpointException("Router::getServerProxy returned a null proxy.");
+        }
+        serverProxy = serverProxy.ice_router(null); // The server proxy cannot be routed.
```
[the remark this answers is not in the document]
Comment (pepone) [reply js/src/Ice/RouterInfo.js:66]: You are right, we do this everywhere I will fix it.

**Pull request merged**